### ISSUE TITLE
Implement TOML difference

### ISSUE THREAD

**Issue by chshersh** [kowainik/tomland#186]
Currently if you implement `Codec` for file that contains more fields that you need, `tomland` will use only those fields that you specified. However, sometimes it's desired to tell about redundant or unused fields. In JSON it is completely okay to ignore some fields, but in TOML static configuration having some fields in the config may lead to frustration in the future. It would be nice to have some way of telling which fields were not used.

**Comment by willbasky** [comment]
Should it be cli command or stderr message after decoding?

**Comment by chshersh** [comment]
@willbasky Under one issue I proposed the following data type as a result of decoding TOML:

```haskell
data Result a
    = Failure String
    | Warning [DecodeWarning] a
    | Success a
```

I believe this should be a separate function(s) in some experimental `Toml.Diff` module. And design for this feature is a subject for discussion :thinking:

**Comment by willbasky** [comment]
Ok. Then for this issue it will be within result message `| Warning [DecodeWarning] a`

And for other issue it will be separate function. And maybe we will find some other useful functions to be invented.

**Comment by willbasky** [comment]
Now we have `DecodeException`. 
`decode` returns `Either DecodeException a`
Should it returns `Result a`? where 
```
data Result a
    = DecodeException
    | Warning [DecodeWarning] a
    | Success a
```

**Comment by chshersh** [comment]
@willbasky In future we might want to change `decode` to return `Result`. But only after this experimental approach will be battle-tested.

**Comment by willbasky** [comment]
So, now, should it be in separate module. Some decode-like function that returns `Result a`?

**Comment by chshersh** [comment]
@willbasky Exactly

**Comment by chshersh** [comment]
The function is starting to become more useful. So here is the roadmap for implementing the feature:

- [ ] Implement the `tomlDiff :: TOML -> TOML -> TOML` function. This is probably the most difficult part of this issue.
- [ ] Add one more constructor to `TomlDecodeError`, something like `NotExactDecode TOML`.
- [ ] Implement the `exactDecode` function with the following logic:
    1. Run `decode` to get the value.
    2. Convert the resulting value back to TOML.
    3. Find diff between the original TOML and the resulting TOML.
    4. If the diff is not empty, fail with `NotExactDecode TOML` with diff inside error message
- [ ] Write some tests :slightly_smiling_face:
